src/Raven.Server/Documents/Queries/IndexQueryServerSide.cs

```diff
@@ -302,22 +301,21 @@ void SetupPagingFromQueryMetadata()
             {
                 if (result.Metadata.Query.Offset != null)
                 {
-                    var offset = (int)QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.Offset, 0);
+                    var offset = QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.Offset, 0);
                     result.Offset = offset;
                     result.Start = start + offset;
                 }
 
                 if (result.Metadata.Query.Limit != null)
                 {
-                    pageSize = (int)QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.Limit, int.MaxValue);
+                    pageSize = Math.Min(int.MaxValue, QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.Limit, int.MaxValue));
                     result.Limit = pageSize;
                     result.PageSize = Math.Min(result.PageSize, pageSize);
                 }
 
                 if (result.Metadata.Query.FilterLimit != null)
                 {
-                    result.FilterLimit =
-                        (int)QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.FilterLimit, int.MaxValue);
+                    result.FilterLimit = QueryBuilderHelper.GetLongValue(result.Metadata.Query, result.Metadata, result.QueryParameters, result.Metadata.Query.FilterLimit, int.MaxValue);
```

Comment:
Shouldn't be limited to `int.Max` as well?

Reply:
yes